Optimize ST_EXTENT_AGG for geo_shape and cartesian_shape

craigtaverner opened the pull request:
Support for `ST_EXTENT_AGG` was added in https://github.com/elastic/elasticsearch/pull/118829, and then partially optimized in https://github.com/elastic/elasticsearch/pull/118829. This optimization worked only for cartesian_shape fields, and worked by extracting the Extent from the doc-values and re-encoding it as a WKB `BBOX` geometry. This does not work for geo_shape, where we need to retain all 6 integers stored in the doc-values, in order to perform the datelline choice only at reduce time during the final phase of the aggregation.

Since both geo_shape and cartesian_shape perform the aggregations using integers, and the original Extent values in the doc-values are integers, this PR expands the previous optimization by:
* Saving all Extent values into a multi-valued field in an IntBlock for both cartesian_shape and geo_shape
* Simplifying the logic around merging intermediate states for all cases (geo/cartesian and grouped and non-grouped aggs)
* Widening test cases for testing more combinations of aggregations and types, and fixing a few bugs found

craigtaverner committed:
Support extracting extent from doc values as int[] for both cartesian and geo shapes

craigtaverner committed:
Merge remote-tracking branch 'origin/main' into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Update docs/changelog/119889.yaml

craigtaverner committed:
Fix changelog

craigtaverner committed:
Added capability to block mixed-cluster tests which fail

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Revert "Added capability to block mixed-cluster tests which fail"

This reverts commit 898d1ba262984909914d82e27f39243c6d687024.

craigtaverner committed:
Change ST_EXTENT_AGG tests on airport_city_boundary to avoid multi-cluster issues

Currently the fact that this index is also used for ENRICH means the test infrastructure duplicates the index across the cluster, which leads to incorrect results in COUNT stats (double the rows). We avoid this by using airports index only if we want to get counts, and ENRICH against airport boundaries to get the geo_shapes.

craigtaverner committed:
Add back non-ENRICH versions to be sure we have multi-shape aggs

These versions cannot use COUNT() due to the issue with indexes that are also used in enrich policies being duplicated across multi-cluster IT environments.

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Bring back capability because the mixed-cluster tests failed on the change in meaning

Of some int values in the block, particularly those that mean unset or infinity.

craigtaverner committed:
Reorder parameters in geo_point/geo_shape aggregators

We had many different parameter orderings, but want to stardize on just two:
* For cartesian we use Rectangle class order (minX, maxX, maxY, minY)
* For geo we use Extent class order (top, bottom, negLeft, negRight, posLeft, posRight)

Since the PR already changed the meaning of the invalid/infinite values of the intermediate state integers, it was already not compatible with the previous cluster versions. We disabled mixed-cluster testing to prevent errors as a result of that. This leaves us the opportunity to make further changes that are mixed-cluster incompatible, hence the decision to perform this consistency update now.

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

GalLalouche commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:
```diff
@@ -497,6 +511,23 @@ private void combineRawInputForBytesRef(MethodSpec.Builder builder, String block
         builder.addStatement("$T.combine(state, $L.getBytesRef(i, scratch))", declarationType, blockVariable);
     }
 
+    private void startWarningsBlock(MethodSpec.Builder builder) {
```
Since any piece of code which starts with `startWarningBlock` should always end with `endWarningsBlock`, I think this should be enforced, e.g., by passing in a lambda to a template which wraps it with those two functions.

craigtaverner replied:
If this is not enforced, and someone makes a mistake, the generated code will not compile, so we immediately get a failure. I think that is sufficient, and probably the reason this was not done before.

GalLalouche commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/AggregatorImplementer.java:
```diff
@@ -693,4 +716,24 @@ public void assignToVariable(MethodSpec.Builder builder, int offset) {
+
+    private String valueTypeString() {
+        String valueTypeString = TypeName.get(valueTypeMirror()).toString();
+        if (valuesIsArray) {
```
Consider using ternary expression instead.

craigtaverner replied:
Done

GalLalouche commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/GroupingAggregatorImplementer.java:
```diff
@@ -407,70 +422,54 @@ private MethodSpec addRawInputLoop(TypeName groupsType, TypeName valuesType) {
             combineRawInputWithTimestamp(builder, offsetVariable);
         } else if (valueType.isPrimitive() == false) {
-            throw new IllegalArgumentException("second parameter to combine must be a primitive");
+            throw new IllegalArgumentException("second parameter to combine must be a primitive, array or BytesRef: " + valueType);
```
Perhaps rephrase to make the error clearer, e.g., `second parameter to combine must be a primitive, array or BytesRef, but was: " +`

craigtaverner replied:
I prefer to keep the message as short as possible.

GalLalouche commented on x-pack/plugin/esql/compute/gen/src/main/java/org/elasticsearch/compute/gen/GroupingAggregatorImplementer.java:
```diff
@@ -489,6 +488,22 @@ private void combineRawInputForBytesRef(MethodSpec.Builder builder, String block
+        }
+    }
+
+    private void endWarningsBlock(MethodSpec.Builder builder) {
```
Reduce duplication with `AggregatorImplementer`?

craigtaverner replied:
There are actually many functions duplicated between these two classes. This PR does take a step towards making similar functions even more similar, so that such a duplication reduction can be done in the future, but does not take that step, since it seems out of scope and makes the PR much messier. I had to draw the line somewhere.

GalLalouche commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/spatial/SpatialExtentGroupingState.java:
```diff
@@ -66,6 +67,32 @@ public void toIntermediate(Block[] blocks, int offset, IntVector selected, Drive
+            int minY = values[3];
+            add(groupId, minX, maxX, maxY, minY);
+        } else {
+            throw new IllegalArgumentException("Expected 4 or 6 values, got " + values.length);
```
Perhaps also add the array itself to the error for good measure.

craigtaverner replied:
I prefer to keep the error message a little more concise. In the case where the length is not 4 or 6, the meanings of the values will be entirely undefined, and we will need to do a deeper investigation, and so printing the values in the error will not change that. There is also a policy regarding avoiding printing data content in error messages, because it results in logfiles potentially containing data that should not be visible.

GalLalouche commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/spatial/SpatialExtentGroupingStateWrappedLongitudeState.java:
```diff
@@ -102,53 +104,74 @@ public void add(int groupId, SpatialExtentGroupingStateWrappedLongitudeState inS
+     * This optimization is enabled when the field has doc-values and is only used in the ST_EXTENT aggregation.
+     */
+    public void add(int groupId, int[] values) {
+        if (values.length == 6) {
```
Consider using early exit here, i.e., first validating the length, then continuing the rest of the code outside of the `else` block. It reveals intent better.

GalLalouche commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/spatial/SpatialExtentStateWrappedLongitudeState.java:
```diff
@@ -37,44 +44,65 @@ public void close() {}
+     * This optimization is enabled when the field has doc-values and is only used in the ST_EXTENT aggregation.
+     */
+    public void add(int[] values) {
+        if (values.length == 6) {
```
Same here, use early exit.

GalLalouche commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/spatial.csv-spec:
```diff
@@ -547,35 +551,260 @@ BBOX (70.77995480038226, 91.5882289968431, 33.9830909203738, 8.47650992218405)
+
+FROM airports
+| WHERE country == "India"
+| STATS extent = ST_EXTENT_AGG(location), count = COUNT()
```
Out of curiosity, why the added `count`?

craigtaverner replied:
Two reasons, I wanted tests that tested the combination of spatial and non-spatial aggregation functions, and I also wanted to see results with counts == 1 and > 1 because that tests different code paths in the aggregation.

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/aggregate/SpatialExtent.java:
```diff
@@ -110,10 +113,17 @@ public AggregatorFunctionSupplier supplier(List<Integer> inputChannels) {
+            case DataType.GEO_SHAPE -> switch (fieldExtractPreference) {
+                case EXTRACT_SPATIAL_BOUNDS -> new SpatialExtentGeoShapeDocValuesAggregatorFunctionSupplier(inputChannels);
+                case NONE -> new SpatialExtentGeoShapeSourceValuesAggregatorFunctionSupplier(inputChannels);
+                default -> throw new EsqlIllegalArgumentException("Illegal field extract preference: " + fieldExtractPreference);
```
Since this is a very small `enum`, it's better to explicitly match all known variants. That way if a new variant is added, it will fail compilation instead of (maybe) at runtime (here and below).

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/SpatialShapeBoundsExtraction.java:
```diff
@@ -49,9 +50,20 @@
     @Override
     protected PhysicalPlan rule(AggregateExec aggregate, LocalPhysicalOptimizerContext ctx) {
-        var foundAttributes = new HashSet<Attribute>();
+        var foundAttributes = findSpatialShapeBoundsAttributes(aggregate, ctx);
```
Don't use `var` here, as the type isn't clear from the name or expression.

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/SpatialShapeBoundsExtraction.java:
```diff
@@ -84,18 +96,27 @@ protected PhysicalPlan rule(AggregateExec aggregate, LocalPhysicalOptimizerConte
+        return foundAttributes;
+    }
+
+    private PhysicalPlan transformFieldExtractExec(FieldExtractExec fieldExtractExec, Set<Attribute> foundAttributes) {
```
Can be `static`.

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/SpatialShapeBoundsExtraction.java:
```diff
@@ -84,18 +96,27 @@ protected PhysicalPlan rule(AggregateExec aggregate, LocalPhysicalOptimizerConte
+        return fieldExtractExec.withBoundsAttributes(boundsAttributes);
+    }
+
+    private PhysicalPlan transformAggregateExec(AggregateExec agg, Set<Attribute> foundAttributes) {
```
Can be `static`.

GalLalouche commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/physical/local/SpatialShapeBoundsExtraction.java:
```diff
@@ -49,9 +50,20 @@
+    }
 
-        return aggregate.transformDown(UnaryExec.class, exec -> {
+    private Set<Attribute> findSpatialShapeBoundsAttributes(AggregateExec aggregate, LocalPhysicalOptimizerContext ctx) {
```
Can be `static`.

GalLalouche commented on x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/planner/TestPhysicalOperationProviders.java:
```diff
@@ -498,12 +495,123 @@ protected long encode(BytesRef wkb) {
+                for (int c = 0; c < docIndices.getPositionCount(); c++) {
+                    int doc = docIndices.getInt(c);
+                    int count = bytesRefBlock.getValueCount(doc);
+                    int i = bytesRefBlock.getFirstValueIndex(doc);
```
Move this nearer to where it is used?

craigtaverner replied:
I wanted it close to the previous line, so we get the key information we need from the bytesRefBlock in one place.

craigtaverner replied:
I found a way to move it and make it look better.

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner commented on libs/geo/src/main/java/org/elasticsearch/geometry/utils/SpatialEnvelopeVisitor.java:
```diff
@@ -186,82 +196,118 @@ public Rectangle getResult() {
-            } else if (Double.isInfinite(minNegX)) {
-                return new Rectangle(minPosX, maxPosX, maxY, minY);
+            assert Double.isFinite(top);
+            // Due to this data coming through Extent (and aggs that use the same approach), which saves values as integers,
```
The comment is meant to explain why the code works. We could change the code to be more like the comment, but I wanted to minimize code changes, while simultaneously increasing clarity by adding the comment. The key point here is when looking for infinities we have to look at posRight and negLeft.

craigtaverner replied:
I'll delete the comment and change the code to instead do that explicitly.

craigtaverner commented on server/src/main/java/org/elasticsearch/index/mapper/AbstractShapeGeometryFieldMapper.java:
```diff
@@ -119,27 +115,27 @@ public BlockLoader.Block read(BlockLoader.BlockFactory factory, BlockLoader.Docs
-                        );
-                        builder.appendBytesRef(new BytesRef(WellKnownBinary.toWKB(rectangle, ByteOrder.LITTLE_ENDIAN)));
+                        // We store the 6 values as a single multi-valued field, in the same order as the fields in the Extent class
+                        // This requires that consumers also know the meaning of the values, which they can learn from the Extent class
```
I cannot think of a clearer way of saying this. Can you make a suggestion?

craigtaverner replied:
Is it the word 'Consumer'? I consider this code the producer (produces an int[6] array). Any consumer that should read that array needs to know what each of the 6 values means. For example, it needs to know that a[1] means the bottom latitude (minimum y value).

GalLalouche replied:
I think it's redundant. It is already stated that the 6 values are stored in same order as the `Extent` class, what is there to "learn"? Perhaps it could be rephrased as: "See that class for the meaning of these values".

craigtaverner replied:
I see now that you were asking for less verbocity, since my two comment lines overlap quite a bit. I've deleted the second line, as I agree it is excessive.

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Code review updates

craigtaverner committed:
Merge branch 'st_extent_optimize_geoshape' of github.com:craigtaverner/elasticsearch into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
One more code review comment

craigtaverner committed:
Merge branch 'st_extent_optimize_geoshape' of github.com:craigtaverner/elasticsearch into st_extent_optimize_geoshape

craigtaverner committed:
Two more code review comments

craigtaverner committed:
Support optimization for cartesian shapes

Only saves 4 ints to the IntBlock from extents. This moves the conversion of 6->4 from the aggregator to the blockloader, which is a slight memory efficiency improvement.

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Updated SpatialGeometryFieldMapperTests and identified bug

We left the bug as a TODO for the next PR, so at least the primary optimization can get merged for tech-preview in 8.18.

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'st_extent_optimize_geoshape' of github.com:craigtaverner/elasticsearch into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Support missing groupIds in intermediate state creation

craigtaverner committed:
Merge branch 'st_extent_optimize_geoshape' of github.com:craigtaverner/elasticsearch into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Added more cartesian tests with larger result sets

This is to increase the chance that the same failure seen for geo_shape in serverless would also occur with cartesian_shape in serverless.

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

craigtaverner committed:
Merge branch 'main' into st_extent_optimize_geoshape

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.x  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 119889`

craigtaverner commented:
Backports done manually:
* 8.x (8.19) in https://github.com/elastic/elasticsearch/pull/122276
* 8.18.1 in https://github.com/elastic/elasticsearch/pull/122420
